### DISCUSSION TITLE
Few metadata related questions (and a missing option from main repo?)

### DISCUSSION THREAD

**Discussion by TKS4** [vladmandic/sdnext#818] [Q&A] [answered]
I've attached an image for clarity,

Maybe I'm going about this the wrong way, but the way I view metadata for images within the UI, I use the Process Image tab.

In the Process Image metadata view, there's two categories with one category having two sub categories. There's Parameters, and then there's Full Metadata, The Parameters tab is nicely formatted, but the Full Metadata has "parameters" and "UserComment" sections that are identical.

What exactly is the point of these repeats given that it seems they all present all the the metadata usually made available, and how would one even edit the "UserComment" section (given that it implies you could).

Also, I'm wondering - where did the option to quell certain metadata from generations? The options to leave out things like Seed and/or Model information of the generated image? 

Lastly, would it be possible to create one more metadata field, and that's of the Sigma Noise slider's value? I'm running a few tests, but not having this info readily apparent is just a little tough. (I realize I should make an issue for this potentially, and also of the LyCORIS quicksetting tab not having a NONE option like the LORA quicksetting tab does)

![STABLEMETA](https://user-images.githubusercontent.com/19920323/236894409-d9d19ec1-7bee-4c9c-9378-2598ab3971bb.jpg)

**Comment by vladmandic** [comment]
Parameters is what has been succesfully parsed from full metadata.
Full metadata can be anything - for example, if you edit an image in Photoshop, it will add its metadata.
That functionality did not exist in A1111.

Regarding adding additional fields, adding them to metadata is easy - parsing them on load then updating UI settings based on that is a nightmare. So no unless its a high value item requested by multiple users.

**Reply by TKS4** [reply]
Ah I get it now. The top field is what the UI can cleanly parse from all the metadata, and the rest is the actual "full" metadata in some readable form I presume? And what about "UserComment" what's that all about - or is that the section that if I edit in photoshop, it will change?

As for the "easy version" of adding it to metadata, that would be perfect, I don't need any interactive functionality for the metadata field. All I care about is being able to see what it is at all. It doesn't need to be able to carry-over to other functionality (like "Load from Last Prompt" or things of that nature). I'd be 100% satisfied if I could simply see it at all anywhere (be it the parser, the raw metadata read from Full Metadata).

Oh and idk if you missed it, but do you have any comments on toggles for certain data being embedded like the Seed or Hash or Model (or whatever else) that the 1111 UI had in it's settings option? Would that be difficult to bring back, and would you recommend I make an issue for it so you can keep track of it?

**Comment by vladmandic** [comment, marked as answer]
Re: UserComments - that is actual EXIF section where we write parameters - so if you open it in Photoshop or wherever, you'll see it.

Re: Toggles - I don't want to have too many toggles. Again, using Photoshop as an example, it doesn't give you a choice what it writes in metadata.

> I don't need any interactive functionality for the metadata field. All I care about is being able to see what it is at all

The problem is that you may not, but trust me as soon as I add it, someone will create an issue how that new field is not properly interpreted by UI (already happened recently - when I added `clip-skip`).

**Reply by TKS4** [reply]
Wasn't the clip SKIP update more serious? Where it can be used to carry-over information when you "Load of Last Prompt" type?

The reason I ask is because I actually have no idea how I'd go about determining what the Sigma Noise slider value is in any way. In fact, none of this needs to be handled by you now that I think about it - do you perhaps know of any relevant add-on, or even a totally separate program if need be, to achieve any of these sorts of things I talked about? Mainly being able to read the Sigma Noise value, and being able to delete metadata fields themselves?

I guess using a Hex editor is possible, but I don't know of a way I'd be able to batch this sort of process (of for instance: removing every seed meta data field from a set of images I have) without going in, one by one, tediously doing it.

As for "not adding too many toggles", the toggles already exist in the main repo. Surprised if you would have to add anything yourself, though I don't know, maybe your fork has changed a lot, but this strikes me as something relatively simple (to skip adding model, model hash, and seed info to generated images)?

**Comment by vladmandic** [comment]
ok, can someone make a list of all options that would be useful to add to metadata? so its not a piece meal (add sigma noise now add one more next week, etc.)?

re: toggles - one of my core goals is to remove as many unnecessary switches/toggles/options as possible and have a better built-in intelligence. i've already removed 80% of command line switches and also removed over 30 options in settings - and you havent even noticed 28 other than 2 you just mentioned :)
and thats even with tons of newly added new functionality.

**Reply by TKS4** [reply]
Yeah I noticed for sure some 1111 UI changes, like removing that weird watermark toggle that to my knowledge was vaporware that never got implemented (kinda funny that UI elements would be completed for a feature that would never even be released, seems a bit backward to me lol). It's one of the great things I think your fork has done.

I could live without the Sigma Noise being parsed or being a metadata field tbh (that feature would be a creature comfort for me personally for tests that will eventually conclude anyway), but the toggles in the Settings Image tab for Model, Model Hash, and Seed metadata not being included in generated images would be something useful to perpetuity for blind tests I'd be running with users.

One last thing since it was a bit tangential and I think we glanced over it. Would adding that "None" option to the LyCORIS quicksettings dropdown be something too difficult to implement?

**Reply by vladmandic** [reply]
re: sigma noise - i can add that, but like i said, i'd like to get a more comprehensive list of what else should be added so i do it once and not revisit every week.

re: lycoris - pretty difficult primarily because i'm including original lycoris as-is - as soon as i modify a single line, then its out-of-sync and auto-update will not work.

**Reply by TKS4** [reply]
Alrighty, then that settles it I got all the answers I need, thanks for your patience Vlad. Nevermind on the Sigma Noise thing, since no one else is really raising all that much of a fuss over particulars like that. And nevermind the LyCORIS quick-setting, I think just typing None works just as well.

Though if you ever change your mind about adding the Model Name, Model Hash, and Seed toggles back into the Image Settings tab I'll be thrilled to see those, I'll go see if there's a smarter way of dealing with this other than busting out a hex editor.

Thanks again,

P.S. Surprise Surprise, RedBull on top again huh -.- Hope you enjoyed the race.

**Comment by xedis** [comment]
Glad I found this post! 

So one I'd like to put my vote in for the additional PNG metadata. If that isn't going to happen though, if you could point me to the section of code & commit this is live in so I can snag the code for my own purposes I'd forever be grateful. 

Just looking for a starting point to script some solution that I can run on the fly to populate/create all the relevant values for indexing and tagging within my db and across applications as seamlessly as possible. PNG exif data seems to be the most uniform and flexible. Currently sitting on 200k images and I'd like to be able to filter by Aesthetic Score within Digikam since webui cannot handle that many lol

Been beating my head against the fact it was present for like a week and then gone for about 2 weeks now when I noticed it present in all my gens between 4-27 and 5-8


(https://github.com/vladmandic/automatic/assets/4144634/20a31e09-f397-4e91-a200-fb051c979382)

```
(base) ➜  automatic ~/Image-ExifTool-12.62/exiftool ~/Desktop/exifbug/20230427-00066-1042377740-.1girl-jibril-ageless-Egyptian-beautiful-unconscious-fully-flesh-face-long-fangs-sharp-canine-teeth-bio-punk-android-full.png
ExifTool Version Number         : 12.62
File Name                       : 20230427-00066-1042377740-.1girl-jibril-ageless-Egyptian-beautiful-unconscious-fully-flesh-face-long-fangs-sharp-canine-teeth-bio-punk-android-full.png
Directory                       : /home/drekan/Desktop/exifbug
File Size                       : 553 kB
File Modification Date/Time     : 2023:04:27 19:04:01-07:00
File Access Date/Time           : 2023:06:15 12:33:51-07:00
File Inode Change Date/Time     : 2023:06:11 00:59:43-07:00
File Permissions                : -rw-rw-rw-
File Type                       : PNG
File Type Extension             : png
MIME Type                       : image/png
Image Width                     : 640
Image Height                    : 960
Bit Depth                       : 8
Color Type                      : RGB
Compression                     : Deflate/Inflate
Filter                          : Adaptive
Interlace                       : Noninterlaced
Parameters                      : .1girl jibril, ageless Egyptian beautiful,  ,unconscious fully flesh face,long fangs,sharp canine teeth,bio-punk,android,, full body shot from the ground, inhuman shaped skull, ,extreme detail,,eyes rolled back,young,s, head tilted to side,neon green hair, realistic, real, perfect claws,jibril.Negative prompt: ugly,worst quality, low quality,sagging, old,text,watermark,logo.Steps: 10, Sampler: DPM++ 2M Karras, CFG scale: 7.5, Seed: 1042377740, Size: 512x768, Model hash: e1c2a8083a, Model: __greenistheneworange_, VAE: NAI_animefull-final.vae, Denoising strength: 0.7, Clip skip: 2, Token merging ratio hr: 0.5, ControlNet Enabled: True, ControlNet Preprocessor: canny, ControlNet Model: control_sd15_canny [fef5e48e], ControlNet Weight: 1, ControlNet Starting Step: 0, ControlNet Ending Step: 1, ControlNet Resize Mode: Crop and Resize, ControlNet Pixel Perfect: False, ControlNet Control Mode: Balanced, ControlNet Preprocessor Parameters: "(64, 64, 64)", Hires upscale: 1.25, Hires steps: 14, Hires upscaler: R-ESRGAN General 4xV3, Score: 4.8
Aesthetic score                 : 5.6
Sampler                         : DPM++ 2M Karras
Cfg scale                       : 7.5
Sd model hash                   : e1c2a8083a
Seed                            : 1042377740
Hash                            : 4855bcf2daeab371559b0ff7c02cec63
Image Size                      : 640x960
Megapixels                      : 0.614
```
(young was basically required to not get a mummy, irrelevant but just wanted to throw that out there.)

Annnnnd the day it stopped being present *sobs*

(https://github.com/vladmandic/automatic/assets/4144634/22dbe376-f1c6-423f-a4e3-c9e0bc0f0a80)
```
(base) ➜  automatic ~/Image-ExifTool-12.62/exiftool ~/Desktop/exifbug/20230509-20235524-2143934611-1girl-solo-official-art-unity-8k-wallpaper-ultra-detailed-beautiful-and-aesthetic-masterpiece-best-quality-realistic-z.png
ExifTool Version Number         : 12.62
File Name                       : 20230509-20235524-2143934611-1girl-solo-official-art-unity-8k-wallpaper-ultra-detailed-beautiful-and-aesthetic-masterpiece-best-quality-realistic-z.png
Directory                       : /home/drekan/Desktop/exifbug
File Size                       : 2.4 MB
File Modification Date/Time     : 2023:05:09 04:01:44-07:00
File Access Date/Time           : 2023:06:15 12:47:31-07:00
File Inode Change Date/Time     : 2023:06:15 12:47:31-07:00
File Permissions                : -rw-rw-r--
File Type                       : PNG
File Type Extension             : png
MIME Type                       : image/png
Image Width                     : 1024
Image Height                    : 1280
Bit Depth                       : 8
Color Type                      : RGB
Compression                     : Deflate/Inflate
Filter                          : Adaptive
Interlace                       : Noninterlaced
Parameters                      : 1girl, solo, official art, unity 8k wallpaper, ultra detailed, beautiful and aesthetic, masterpiece, best quality, realistic, (zentangle, mandala, tangle, entangle), (fractal art), (silk, spider web, centipede), extremely detailed, dynamic angle, the most beautiful form of chaos, elegant, a brutalist designed, vivid colours, romanticism, zhongfenghua <lora:zhongfenghuaxiyou-000018:0.6> grey mist,  <lora:LowRA:0.4>.Negative prompt: nsfw, ng_deepnegative_v1_75t, (badhandv4:1.2), (worst quality:2), (low quality:2), (normal quality:2), lowres, bad anatomy, bad hands, normal quality, ((monochrome)), ((grayscale)) watermark.Steps: 20, Sampler: Heun, CFG scale: 26, Seed: 2143934611, Size: 1024x1280, Model hash: f1a245e4ac, Model: unrule_, VAE: vae-ft-mse-840000-ema-pruned, Variation seed: 3930125616, Variation seed strength: 0.47, Clip skip: 2, Token merging ratio hr: 0.5, Hashes: {"vae": "735e4c3a44", "embed:badhandv4": "5e40d722fc", "embed:ng_deepnegative_v1_75t": "54e7e4826d", "lora:LowRA": "348071db54", "model": "f1a245e4ac"}
Image Size                      : 1024x1280
Megapixels                      : 1.3
```

Not sure if you still need the list of what fields would be relevant / desired I can draft up a list if you want?

**Reply by vladmandic** [reply]
> So one I'd like to put my vote in for the additional PNG metadata. If that isn't going to happen though, if you could point me to the section of code & commit this is live in so I can snag the code for my own purposes I'd forever be grateful.

Sure. And one option you did not suggest is to create actual PR and contribute.

> Not sure if you still need the list of what fields would be relevant / desired I can draft up a list if you want?

But I'm not sure what is the ask?
Is it the fact that aesthetic score is missing in infotext? If thats the case, please create issue in that repo as aesthetic score on its own does not exist at all in the core app.
Which other fields are you talking about?

**Reply by TKS4** [reply]
I know I'm not the guy you're conversing with atm. But I was wondering would now be an okay time to ask if there could be that functionality in the Settings tab as a toggle to leave out model name/hash/seed metadata values of generated images O_o

Or are you still mega swarmed >_<

**Reply by vladmandic** [reply]
> I know I'm not the guy you're conversing with atm

That is not a problem, multiple conversations are ok :)

> toggle to leave out model name/hash/seed metadata values of generated images

Then there should be a toggle for any metadata field and that devolves into a total mess.
Standard industry practice is that app writes its metadata and you can use any 3rd party software to strip it.
For example, can you choose which metadata Photoshop puts in an image?

> Or are you still mega swarmed >_<

Always, but never stopped me from looking at requests :)

**Reply by TKS4** [reply]
Thank vlad, but do you think offering basics like those would be okay? Nothing like 3rd party extensions. So simple things like those three I mentioned and I guess other basics like Positive or Negative promps, CFG value, Sampler, and Sampler Step count? I think that should cover 99% of use cases. Or are you mentioning this worry about devolving into something that will need to constantly be updated because once you let one meta-data value be toggled, people are going to be asking for every single other field to have it's own toggle?

Photoshop btw could work fine, but imagine when you have to do this for a batch of images you want to send out to people for blind tests for example. Doing each image one by one is a nightmare (I don't know what program can be used to "CRTL+F" and mass delete certain metadata fields). 

The only reason I asked for this functionality originally is because it was sorely missed in the original repo, I don't think I've seen anyone ask for an expansion of the fields that can be left out in the generation of images (nor tbh, any commentary on the 3 fields that were on offer from who knows how long ago). Do you have any advice on how this can be done efficiently to cull a ton of generations? That's my biggest problem - I just don't know how this can be done.

**Reply by vladmandic** [reply]
i can understand if artists wants to have no metadata shared as its his proprietary value.
but hiding some metadata and while including other? what's the value, what's the use-case? 
e.g., i'm going to share my photo with metadata, just not prompt. or just not model. it maybe nice-to-have and would be used by few people, but what is real value of that. metadata is either complete or its not.

i'm totally open to add more metadata fields. and i'm open to remove it all. but not have per-field on/off.

**Reply by TKS4** [reply]
A toggle for all metadata to not appear is also fine (didn't know if it was more work or not). The use case I have is I sometimes send out images to blind test between new model merges or generations that use certain negative embeddings (which can dilute certain merges, while not having as strong of an effect as others), if I don't purge the metadata, then it's simple to see which batch of images comes from what merge/set of prompts. When you want to be sure no one can bypass the blind test, then you have to get rid of the metadata.

So sure, a global off toggle would also be great, the only reason I didn't ask for it prior, was as I mentioned in my opening, didn't want it to be too much work (as I thought an import of the old 1111 code might've been easier for you). But if you say a metadata on/off toggle is simpler, then that would be a nice update. <3

**Reply by xedis** [reply]
Sorry had an emergency work call....

Ah right! I left out my past week of floundering around since I didn't end up submitting the issue I drafted twice while I was trying to figure it out where the root source of the additional fields was even to start with. Once i identified the dates & i tried checking out the commits i thought might be it based on the the images.py file being modified. Alas no dice. I'd all but given up thinking it a fever dream until i stumbled here in the middle of my workday. 

During that rabbit hole I dug into the documentation for ExifTool and Exiv2. Also jumped through the digikam code to make sure it wasn't one of it's many ML addons.
> 
> Sure. And one option you did not suggest is to create actual PR and contribute.
>

Absolutely, and 100% if I manage to get something functional going, a PR would definitely be something I would be looking at submitting back into the project. 

> But I'm not sure what is the ask? Is it the fact that aesthetic score is missing in infotext? If thats the case, please create issue in that repo as aesthetic score on its own does not exist at all in the core app. Which other fields are you talking about?

I was just hoping for the keys/values in Parameters being applied as PNG:exif in general. Right now it's all just being dropped in the Parameters/UserComment section as a string, one that has some inconsistencies with formatting to say the least. I understand,thanks to one of the closed issues(dont have it on hand to reference :( ), that sometimes extensions append the field, thus not something likely easily wrangled.

Aesthetic score was one part, I don't think it was ran against this specific image so that's fine, apologies, bad example. 

The fields:

|Key | 
| --- | 
Aesthetic score|           
Sampler |                
Cfg scale | 
|Sd model hash| 
|Seed |        
|Hash |

Other additional entries I could see being useful depending on size constraints.(constraints,I believe were mentioned in one of the issues I found will looking into the missing entries, also don't have on hand.... :( ) 

| Key |
| --- |
|prompt| 
|Var seed |
|Var seed % |
|Hi-res fix    | 
|Upscaler     |

IF i did submit a PR I will also think on how to elegantly layout the settings while I'm at it, so as to not be a hotmes ;) 

My use case is simply just organization and sorting of my ever increasing pile of images. Right now the front end for indexing said metadata fields exists in digikam when it reads the images and inserts them into the db. But only to the granularity I'm looking for when they exist as independent entries. 

TL;DR: I was hunting for an easy way to index/tag/categorize/my images, as webui is the first to write to the Parameters value it's not convoluted by extension's additons so it makes a solid base.

My workflow is currently Gen -> Mass Rename every few days using a script -> import into digikam -> digkam inserts into db -> build albums based on various variables. 
I broke the chain oops.

**Reply by vladmandic** [reply]
you have to remember that whatever fields you populate, they have to be portable between different image formats - png, jpg, webp, etc.

and as much as i hate current way of storing everything as a flat string, that option cannot go away as i have to maintain compatibility with previously generated images and/or images generated in other a1111 repos. so whatever this implemenation does, it has to be on top of current implementation, not replacing it.

**Reply by xedis** [reply]
>so whatever this implemenation does, it has to be on top of current implementation, not replacing it.

Definitely, it would basically be redundant to some extent, but this also serves a secondary purpose I suppose..... in the fact that applying them as separate keys and values as well would protect them from modifications to the UserComment field unless specifically desired to be wiped. May also make restoring them easier potentially, dunno would require some tests. 

>you have to remember that whatever fields you populate, they have to be portable between different image formats - png, jpg, webp, etc.

Those fields are non-standard on the format level, brand spankin new, so I would need to research how flexible each format is, PNG is pretty much a free for all so long as you match the core structure from what I've seen.

**Reply by vladmandic** [reply]
> Definitely, it would basically be redundant to some extent

good. and yes, it would - but i'm ok with that.

> Those fields are non-standard on the format level, brand spankin new, so I would need to research how flexible each format is, PNG is pretty much a free for all so long as you match the core structure from what I've seen.

that is true. but if we're going to use alternative exif sections, i'd like to use sections which are not deeply burried (even if they are supported), but also visible using common tools - for example, what can i view in windows explorer. just trying to get maximum value.
